### ISSUE TITLE
For fine-tuning which factors influence iteration time?

### ISSUE THREAD

**Issue by ErfolgreichCharismatisch** [NVIDIA/tacotron2#505]
For fine-tuning which factors of hparams and text-audio-pairs increase iteration time?

**Comment by CookiePPP** [comment]
![image](https://user-images.githubusercontent.com/42448678/126316922-4ce18a7c-88b2-4816-a162-216d40e6b8e5.png)
https://lambdalabs.com/gpu-benchmarks
Check the tacotron2 tab for the effect of different hardware.

---

For decreasing iteration time with hparams, the main thing is `n_frames_per_step` which is stuck at 1 in this repo. 😅 
https://github.com/NVIDIA/tacotron2/blob/185cd24e046cc1304b4f8e564734d2498c6e2e6f/hparams.py#L56

---

For text-audio-pairs, the only thing that can improve iteration time is using shorter audio files, but that will also make your model worse in most cases so not recommended.
I would recommend removing any files that are above 10 seconds or something like that though, just to reduce the chance or suddenly spiking VRAM usage in the middle of the run.

**Comment by ErfolgreichCharismatisch** [comment]
When fine-tuning and having reached 8k steps with a different language and additional characters and inference is not what is written - without Max encoder steps reached -  what can you conclude from that?

Will a customly trained waveglow make it understandable or is it just for making it clearer, so if you cannot understand what is said using the tacotron model, you also won't understand it if you trained waveglow. 

In other words is the tacotron model the be all and end all or is it only half and what matters is how a custom waveglow model sounds with it?

In other words, is waveglow training necessary or luxury? Does that change when using new characters and a different language?

**Comment by CookiePPP** [comment]
@ErfolgreichCharismatisch 
WaveGlow will not change the words spoken. WaveGlow is used to improve audio quality, but it will not help you if you cannot understand what is being said.

Can you describe how it's failing and upload some screenshots from tensorboard of the alignment and predicted spectrogram?

**Comment by ErfolgreichCharismatisch** [comment]
> @ErfolgreichCharismatisch
> 
> Can you describe how it's failing and upload some screenshots from tensorboard of the alignment and predicted spectrogram?

Yes. For instance when entering "Hallo. Wie geht es dir?" German for "Hello, how are you" if you are not german(most are here), I get a reply along the lines of "Hmm hm hmmmm hmm". It is clearly the voice actor, but he cannot be understood.
![Webaufnahme_14-7-2021_114126_colab research google com](https://user-images.githubusercontent.com/18123801/126567697-75f8e9e9-93ee-4641-8d01-da99495fe0fc.jpeg)
![Webaufnahme_22-7-2021_0216_colab research google com](https://user-images.githubusercontent.com/18123801/126567698-32232748-27a1-4fc6-8bad-0c0b08fa8679.jpeg)
![Webaufnahme_22-7-2021_02122_colab research google com](https://user-images.githubusercontent.com/18123801/126567700-cdd406c8-6f08-41c7-b908-af85b7feb8e5.jpeg)
![Webaufnahme_22-7-2021_02150_colab research google com](https://user-images.githubusercontent.com/18123801/126567701-01406039-7df9-4f88-b50e-6bed82f79d0b.jpeg)

**Comment by CookiePPP** [comment]
@ErfolgreichCharismatisch 
This looks very good!
Can you upload some more alignment pictures at 2000, 4000, 6000 and so on?
Right now I think all you need is more training time or maybe more data, everything I can see here suggests the model is simply early into learning and will eventually stabilize.

**Comment by ErfolgreichCharismatisch** [comment]
2400
![2400](https://user-images.githubusercontent.com/18123801/126599324-5718d82f-c7f5-40f0-862c-2592ca2df41c.png)
4000
![4000](https://user-images.githubusercontent.com/18123801/126599328-9d59ef23-cbe2-4569-b288-73c35447f2d5.png)
7600
![7600](https://user-images.githubusercontent.com/18123801/126599329-a1cba111-1c44-44ca-9c8e-9c0910b5b1d2.png)
9400
![9400](https://user-images.githubusercontent.com/18123801/126599330-35bf5114-a319-4e93-ab33-3d238021e246.png)
9600
![9600](https://user-images.githubusercontent.com/18123801/126599331-9494331c-573f-4c4d-8d30-419017363035.png)
![wa](https://user-images.githubusercontent.com/18123801/126600734-d5d825db-97a4-4486-9fc8-401ad9cc507f.jpeg)

**Comment by ErfolgreichCharismatisch** [comment]
Given that I am at 10k steps and 20k are roughly the goal, shouldn't it be a rough but understandable voice?

**Comment by CookiePPP** [comment]
@ErfolgreichCharismatisch 
The model is very sensitive to failure. If any part of the audio fails to generate then most of the audio will fail.
You need the model well trained before you will hear good results.

**Comment by ErfolgreichCharismatisch** [comment]
> @ErfolgreichCharismatisch
> The model is very sensitive to failure. If any part of the audio fails to generate then most of the audio will fail.
> You need the model well trained before you will hear good results.

So the audio text files must be good but I can only see if they are after 20k iterations?
Are you aware of a quicker way?

**Comment by CookiePPP** [comment]
@ErfolgreichCharismatisch 
For a new language, I think it would be faster to let this model train than try and use other methods of checking.

**Comment by CookiePPP** [comment]
![image](https://user-images.githubusercontent.com/42448678/126728968-68f9f36b-3c7c-4ae9-b747-170511bc3ab6.png)

---

Also, if your audio files and text were poor then you would not see a diagonal line appearing in the alignments. It would always look like the right of this image.
I think your model has a good chance of working (more than 80% chance of working), so I think it is perfectly fine to let it continue running.

**Comment by ErfolgreichCharismatisch** [comment]
I figured that some of my audio files were in fact poor, about 10 in 2000 files. But that might have been too much.
Currently training from a warm start again and waiting for gpu on colab...
Speaking of which, do you know a free service that offers gpu usage?

**Comment by CookiePPP** [comment]
> Speaking of which, do you know a free service that offers gpu usage?

I only know of Colab and the free $300 trial on GCP

**Comment by ErfolgreichCharismatisch** [comment]
I fixed all audio files, have them start at the beginning and pad 100ms silence to the end of each.
They are normalized 16 bit audio. 22050 Hz.
I added extra characters to symbols.py.
I added
```
def german_cleaners(text):
  text = GermanTransliterate(replace={';': ',', ':': ' '}, sep_abbreviation=' -- ').transliterate(text)
  return text
```
to cleaners.py and `text_cleaners=['german_cleaners']` to hparams of both inference and training.

I am at step 2000 now.

I still get "Warning! Reached max decoder steps" on inference.

![Webaufnahme_26-7-2021_133225_colab research google com](https://user-images.githubusercontent.com/18123801/126982209-37d064fb-e78d-4bae-babd-25e7420c51f3.jpeg)
![Webaufnahme_26-7-2021_133212_colab research google com](https://user-images.githubusercontent.com/18123801/126982214-2e48553b-0cfa-4db0-8636-97fb5d37e9a6.jpeg)
![Webaufnahme_26-7-2021_133136_colab research google com](https://user-images.githubusercontent.com/18123801/126982219-35800857-2860-42f8-99da-7ab1a88651ae.jpeg)
![Webaufnahme_26-7-2021_133013_colab research google com](https://user-images.githubusercontent.com/18123801/126982220-eb806bca-413b-4d76-b76a-0a599ed483f7.jpeg)

The reason this is a problem is that when not altering symbols.py, it seems to convert already at 200 steps.

I run into gpus not available on colab because trial and error took much time.

Ideas on how to proceed?
